source/common/router/header_parser_utils.cc

```diff
@@ -0,0 +1,82 @@
+#include <regex>
+#include <string>
+
+#include "source/common/common/assert.h"
+#include "source/common/json/json_loader.h"
+#include "source/common/router/header_parser.h"
+
+namespace Envoy {
+namespace Router {
+
+// Related to issue 20389. Header formatters are parsed and processed by formatters defined in
+// source/common/formatter/substitution_formatter.cc. For backwards compatibility UPSTREAM_METADATA
+// and UPSTREAM_METADATA format must be changed. Those formatters used to take a JSON format like
+// UPSTREAM_METADATA(["a", "b"]) and substitution formatters use UPSTREAM_METADATA(a:b) format.
+// This translator translates UPSTREAM_METADATA and DYNAMIC_METADATA from JSON format to colon
+// format.
+// TODO(cpakulski): Eventually JSON format should be deprecated in favor of colon format.
+std::string HeaderParser::translateMetadataFormat(const std::string& header_value) {
+  const std::regex command_w_args_regex(
+      R"EOF(%(UPSTREAM|DYNAMIC)_METADATA\(\s*(\[(?:.|\r?\n)+?\]\s*)\)%)EOF");
```

Comment:
drive-by after-the fact comment.

I'm not sure if this hits the data plane but it would be trivial to use CONSTRUCT_ON_FIRST_USE to compile this regex only once, rather than every time it's called.

Less trivial might be converting it to RE2 for improved speed, or implementing the above regex using direct code. Not sure how important this is.

Reply:
sorry it's obvious from TRY_ASSERT_MAIN_THREAD this does not happen on the data plane. Nevertheless it does happen during operation on xDS updates. Not sure if it's material relative to all that goes on during one.

Reply:
Thanks for reviewing it. This does not happen at the data plane, but during the configuration time. I will convert it to CONSTRUCT_ON_FIRST_USE. 

This is a translator between legacy format of specifying which metadata should be extracted and put into header and the new format. I plan to deprecate legacy format in about 1 year. 
But, the potential problem is that each configured header formatter is matched against this regex. Not sure how efficient is regex. Maybe it should be preceded by simple 
```
(if header_value.StartsWith("UPSTREAM) || header_value.StartsWith("DYNAMIC")) {
  // match regex.

```
I will check it with microbenchmark.

Reply:
The control-plane can also have issues with being compute-bound.

std::regex is particularly bad even if you don't recompile on every call. It allocates memory during parsing, for example. RE2 is much better, but directly parsing in C++ would likely be faster still. Was there any reason not to use RE2? I thought we had close to eliminated std::regex from the codebase.

Reply:
Thanks for that info. There was no particular reason not to use RE2 - just my knowledge on that topic was limited. Let me benchmark it and rework to either to pure C++ or RE2. I will open a new PR and ask you for review.

> I thought we had close to eliminated std::regex from the codebase.

There are still few files which use std::regex. I had to modify tools/code_format/config.yaml to add a new file with std::regex. Now I know why.

Reply:
Yeah the purpose of allow-lists in that file is to limited the spread of bad practice :)